static/app/components/events/eventTags/eventTagsTree.tsx

```diff
@@ -0,0 +1,202 @@
+    return tree;
+  }
+  // E.g. 'device.model.version'
+  const splitIndex = tag.key.indexOf('.'); // 6
+  const trunk = tag.key.slice(0, splitIndex); // 'device'
+  const branch = tag.key.slice(splitIndex + 1); // 'model.version'
+
+  if (tree[trunk] === undefined) {
+    // We need to define the default as an empty space for alignment
+    tree[trunk] = {value: ' ', subtree: {}};
+  }
+  // Recurse with a pseudo tag, e.g. 'model', to create nesting structure
+  const pseudoTag = {
+    key: branch,
+    value: tag.value,
+  };
+  tree[trunk].subtree = addToTagTree(tree[trunk].subtree, pseudoTag, meta, originalTag);
+  return tree;
+}
+
+interface TagsTreeKeysProps {
+  content: TagTreeContent;
+  tag: string;
+}
+
+function TagTreeKeys({content, tag}: TagsTreeKeysProps) {
+  const subtreeTags = Object.keys(content.subtree);
+  return (
+    <TreeKeyTrunk>
+      <TreeKey>{tag}</TreeKey>
+      {subtreeTags.map((t, i) => (
+        <TreeSubtreeWrapper key={`${tag}-${i}-key`} isLast={i === subtreeTags.length - 1}>
+          <TreeSpacer />
+          <TagTreeKeys tag={t} content={content.subtree[t]} />
+        </TreeSubtreeWrapper>
+      ))}
+    </TreeKeyTrunk>
+  );
+}
+interface TagsTreeValuesProps extends TagsTreeKeysProps {
+  projectId: string;
+  projectSlug: string;
+  streamPath: string;
+}
+
+function TagTreeValues({content, tag, ...props}: TagsTreeValuesProps) {
+  const subtreeTags = Object.keys(content.subtree);
+  const organization = useOrganization();
+  const location = useLocation();
+  const originalTag = content.originalTag;
+  return (
+    <TreeValueTrunk>
+      <TreeValue>
+        {originalTag ? (
+          <EventTagsContent
+            tag={originalTag}
+            organization={organization}
+            query={generateQueryWithTag(
+              {...location.query, referrer: 'event-tags-tree'},
+              originalTag
+            )}
+            meta={content?.meta ?? {}}
+            {...props}
+          />
+        ) : (
+          content.value
+        )}
+      </TreeValue>
+      {subtreeTags.map((t, i) => (
+        <TagTreeValues
+          key={`${tag}-${i}-value`}
+          tag={t}
+          content={content.subtree[t]}
+          {...props}
+        />
+      ))}
+    </TreeValueTrunk>
+  );
+}
+
+interface EventTagsTreeProps {
+  projectId: string;
+  projectSlug: string;
+  streamPath: string;
+  tags: EventTag[];
+  meta?: Record<any, any>;
+}
+
+function EventTagsTree({tags, meta, ...props}: EventTagsTreeProps) {
+  const tagTree = tags.reduce<TagTree>(
+    (tree, tag, i) => addToTagTree(tree, tag, meta?.[i], tag),
+    {}
+  );
+  return (
+    <TreeContainer>
+      <TreeGarden>
+        {Object.keys(tagTree).map((tagKey, i) => (
+          <TreeItem key={`${tagKey}_${i}`}>
+            <TagTreeKeys tag={tagKey} content={tagTree[tagKey]} />
+            <TagTreeValues tag={tagKey} content={tagTree[tagKey]} {...props} />
+          </TreeItem>
```

Comment:
So it's a bit odd to separate keys and values into their own components, but this was done since the values need to be aligned vertically with one another, whereas keys may have more spacer components in any given row, and only need their root tag to be aligned. I used `grid-template-columns: subgrid` to accomplish this.